crates/wasm-mutate/src/mutators/peephole/eggsy/encoder.rs

```diff
@@ -177,14 +106,252 @@ impl Encoder {
+                        types[usize::from(operands[0])] = types[usize::from(operands[0])]
+                            .clone()
+                            .or(encoding_tpe.clone());
+                    }
+                    Lang::LtS(operands)
+                    | Lang::LtU(operands)
+                    | Lang::GtS(operands)
+                    | Lang::GtU(operands)
+                    | Lang::LeS(operands)
+                    | Lang::LeU(operands)
+                    | Lang::GeS(operands)
+                    | Lang::Eq(operands)
+                    | Lang::Ne(operands)
+                    | Lang::GeU(operands) => {
+                        // first check if its fixed
+                        let encoding_tpe = types[usize::from(operands[0])]
+                            .clone()
+                            .or(types[usize::from(operands[1])].clone()); // Last from collected info during Wasm2eterm translation
+                                                                          // Set the type for the children
+                        types[usize::from(operands[1])] = types[usize::from(operands[1])]
+                            .clone()
+                            .or(encoding_tpe.clone());
+                        types[usize::from(operands[0])] =
+                            types[usize::from(operands[0])].clone().or(encoding_tpe);
+                    }
+                    Lang::Popcnt(operands) => {
+                        // if its not inferred from the parent, then is the type of the children
+                        types[idx] = types[idx]
+                            .clone()
+                            .or(types[usize::from(operands[0])].clone())
+                            .or(gettpe(Id::from(idx)));
+                    }
+                    Lang::Unfold(operand) => {
+                        // if its not inferred from the parent, then is the type of the children
+                        types[idx] = types[idx]
+                            .clone()
+                            .or(types[usize::from(*operand)].clone())
+                            .or(gettpe(Id::from(idx)));
+                    }
+                    Lang::Num(_) => {
+                        // If it is not set, then take the information from the egraph
+                        types[idx] = types[idx].clone().or(gettpe(Id::from(idx)));
+                    }
+                    Lang::Arg(_)
+                    | Lang::Drop
+                    | Lang::Rand
+                    | Lang::Undef
+                    | Lang::Extend8S(_)
+                    | Lang::Extend16S(_)
+                    | Lang::Extend32S(_)
+                    | Lang::ExtendI32S(_)
+                    | Lang::ExtendI32U(_)
+                    | Lang::Call(_) => {}
+                }
+            }
+            // TODO, check a better wway to check if the type inferring converge
+            if format!("{:?}", previous_pass) == format!("{:?}", types) {
```

Comment:
Yeah, we shouldn't be (ab)using string formatting for to check if we changed anything.

How about, instead of doing `types[idx] = ...` we define and use this little helper:

```rust
let mut changed = false;

let update = |index, ty| {
    let entry = &mut types[index];
    if entry.is_none() {
        *entry = Some(ty);
         changed = true;
    } else {
        assert_eq!(*entry, ty);
    }
};

// ...

if !changed {
    debug_assert!(types.iter().all(|ty| ty.is_some());
    break;
}
```

Reply:
Done